render_delegate/render_delegate.cpp

```diff
@@ -1221,80 +1239,88 @@ bool HdArnoldRenderDelegate::ShouldSkipIteration(HdRenderIndex* renderIndex, con
     }
+    auto& changeTracker = renderIndex->GetChangeTracker();
+    
     auto skip = false;
     if (bits != HdChangeTracker::Clean) {
         renderIndex->GetChangeTracker().MarkAllRprimsDirty(bits);
         skip = true;
     }
     SdfPath id;
-    // We are first removing materials, to avoid untracking them later.
-    while (_nodeGraphRemovalQueue.try_pop(id)) {
-        _nodeGraphToShapeMap.erase(id);
-    }
-    while (_lightNodeGraphRemovalQueue.try_pop(id)) {
-        _nodeGraphToLightMap.erase(id);
-    }
-    ArnoldNodeGraphChange shapeChange;
-    while (_shapeNodeGraphUntrackQueue.try_pop(shapeChange)) {
-        for (const auto& material : shapeChange.materials) {
-            auto it = _nodeGraphToShapeMap.find(material);
-            // In case it was already removed.
-            if (it != _nodeGraphToShapeMap.end()) {
-                it->second.erase(shapeChange.shape);
-            }
+    auto markPrimDirty = [&](const SdfPath& source) {
+        // Marking a primitive as being dirty. But the function to invoke
+        // depends on the prim type. For now we're checking first if a Rprim
+        // exists with this name, to choose between Rprims and Sprims.
+        if (renderIndex->HasRprim(source)) {
+            // for Rprims we 
+            changeTracker.MarkRprimDirty(source, HdChangeTracker::DirtyMaterialId);
         }
-    }
-    while (_shapeNodeGraphTrackQueue.try_pop(shapeChange)) {
-        for (const auto& material : shapeChange.materials) {
-            auto it = _nodeGraphToShapeMap.find(material);
-            if (it == _nodeGraphToShapeMap.end()) {
-                _nodeGraphToShapeMap.insert({material, {shapeChange.shape}});
-            } else {
-                it->second.insert(shapeChange.shape);
-            }
+        else {
+            changeTracker.MarkSprimDirty(source, HdLight::DirtyParams);
         }
-    }
-    ArnoldNodeGraphChange lightChange;
-    while (_lightNodeGraphUntrackQueue.try_pop(lightChange)) {
-        for (const auto& material : lightChange.materials) {
-            auto it = _nodeGraphToLightMap.find(material);
-            // In case it was already removed.
-            if (it != _nodeGraphToLightMap.end()) {
-                it->second.erase(lightChange.shape);
+    };
+    // First let's process all the dependencies that were removed.
+    // We need to remove it from all our maps, and mark all the 
+    // sources as being dirty, so that they can update their 
+    // new reference properly
+    while (_dependencyRemovalQueue.try_pop(id)) {        
+        auto it = _targetToSourcesMap.find(id);
+        if (it != _targetToSourcesMap.end()) {
+            skip = true; // this requires a render update
+            for (const auto& source : it->second) {
+                // for each source referencing the current target
+                // we need to remove the target from its list
+                auto &sourceIt = _sourceToTargetsMap.find(source);
```

Comment:
This should be a const reference `const auto &sourceIt = _sourceToTargetsMap.find(source);`